Add filtering tasks/task instances by tags

mikaelfr opened the pull request:
This pull request adds a tagging system for tasks/task instances similarly to dag tags. This allows filtering by tags in Task Instance view and Graph view. 

A new table, task_tag, is added with TaskInstance's composite primary key as a foreign key. Although having only dag_id and task_id as foreign keys would be sufficient, SQLite doesn't support a foreign key relationship over just a part of a composite primary key. 

---

Screenshots of added UI functionality:
![Filtering by tags](https://user-images.githubusercontent.com/54189907/85530936-fe49b300-b616-11ea-85ae-48ce496456a6.png)
![Task tags in tree view](https://user-images.githubusercontent.com/54189907/85530928-fc7fef80-b616-11ea-94dc-23f5fc105086.png)
![Tags for each task instance](https://user-images.githubusercontent.com/54189907/85530931-fdb11c80-b616-11ea-89bb-b15afe4ace34.png)
![Tags in graph view](https://user-images.githubusercontent.com/54189907/85530932-fdb11c80-b616-11ea-96d9-cff58ccb65c4.png)
![Highlighting tasks by tags in graph view](https://user-images.githubusercontent.com/54189907/85530933-fe49b300-b616-11ea-9297-524f0d18726f.png)

---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

mikaelfr committed:
Add empty migration for TaskTag table.

mikaelfr committed:
Add empty migration for TaskTag table.

mikaelfr committed:
Write migration for task_tag table.

mikaelfr committed:
Write migration for task_tag table.

mikaelfr committed:
Add tags parameter and member variable to BaseOperator.

mikaelfr committed:
Add tags parameter and member variable to BaseOperator.

mikaelfr committed:
Add TaskTag model and sync task tags with database

mikaelfr committed:
Add TaskTag model and sync task tags with database

mikaelfr committed:
Add TaskTag relationship to TaskInstance and display tags in task instance tooltips

mikaelfr committed:
Add TaskTag relationship to TaskInstance and display tags in task instance tooltips

mikaelfr committed:
Fix typo in TaskInstance

mikaelfr committed:
Fix typo in TaskInstance

mikaelfr committed:
Fix relationship in TaskInstance and display task names in tooltip

mikaelfr committed:
Fix relationship in TaskInstance and display task names in tooltip

mikaelfr committed:
Fix bug in BaseOperator and remove tag update from task

* Use tag names instead of entire tag structure when testing for missing/extra tags.
* Remove tags from refresh from task since BaseOperator has tags as a list of strings whereas TaskInstance has a relationship with TaskTag table.

mikaelfr committed:
Fix bug in BaseOperator and remove tag update from task

* Use tag names instead of entire tag structure when testing for missing/extra tags.
* Remove tags from refresh from task since BaseOperator has tags as a list of strings whereas TaskInstance has a relationship with TaskTag table.

mikaelfr committed:
Use tag_id column in task_instance as a foreign key to task_tag

mikaelfr committed:
Use tag_id column in task_instance as a foreign key to task_tag

mikaelfr committed:
Fix problem with ALTER command in sqlite db

mikaelfr committed:
Fix problem with ALTER command in sqlite db

mikaelfr committed:
Add autoincrement and remove sequence

mikaelfr committed:
Add autoincrement and remove sequence

mikaelfr committed:
Use only tag_id as primary key in task_tag

mikaelfr committed:
Use only tag_id as primary key in task_tag

mikaelfr committed:
Fix batch_op in task_instance migration

mikaelfr committed:
Fix batch_op in task_instance migration

mikaelfr committed:
Fix tag id

mikaelfr committed:
Fix tag id

mikaelfr committed:
Revert to
4f7e760

mikaelfr committed:
Revert to
4f7e760

mikaelfr committed:
Add execution_date to TaskTag and remove set_tags_to_db from BaseOperator

Had to add execution_date to TaskTag since SQLite doesn't support foreign keys over just a part of a composite primary keys.

mikaelfr committed:
Add execution_date to TaskTag and remove set_tags_to_db from BaseOperator

Had to add execution_date to TaskTag since SQLite doesn't support foreign keys over just a part of a composite primary keys.

mikaelfr committed:
Use a set in BaseOperator for tags

mikaelfr committed:
Use a set in BaseOperator for tags

mikaelfr committed:
Add TaskInstance join TaskTag in dag and dagrun

mikaelfr committed:
Add TaskInstance join TaskTag in dag and dagrun

mikaelfr committed:
Remove join from DagRun, add tags to serialization schema

mikaelfr committed:
Remove join from DagRun, add tags to serialization schema

mikaelfr committed:
Move join to end and add isouter=True

mikaelfr committed:
Move join to end and add isouter=True

mikaelfr committed:
Load tags into task instance tooltips

mikaelfr committed:
Load tags into task instance tooltips

mikaelfr committed:
Add tags to tree

Tree uses tags from BaseOperator, so tags are now a list instead of a set there because otherwise they won't be JSON serialized.

mikaelfr committed:
Add tags to tree

Tree uses tags from BaseOperator, so tags are now a list instead of a set there because otherwise they won't be JSON serialized.

mikaelfr committed:
Refresh tags from db and update tags on refresh

mikaelfr committed:
Refresh tags from db and update tags on refresh

mikaelfr committed:
Don't update tags in TaskInstance refresh_from_db

* Causes a crash since tags is a relationship

mikaelfr committed:
Don't update tags in TaskInstance refresh_from_db

* Causes a crash since tags is a relationship

mikaelfr committed:
Run static tests

mikaelfr committed:
Run static tests

mikaelfr committed:
Extend task filtering in graph to also look for matches in tags

mikaelfr committed:
Extend task filtering in graph to also look for matches in tags

mikaelfr committed:
Fix small bugs in tag reducer

mikaelfr committed:
Fix small bugs in tag reducer

mikaelfr committed:
Add tags to TaskInstanceModelView and replace hasMatches with has_matches

mikaelfr committed:
Add tags to TaskInstanceModelView and replace hasMatches with has_matches

mikaelfr committed:
Add contains and not contains filters for tags

mikaelfr committed:
Add contains and not contains filters for tags

mikaelfr committed:
Pass datamodel instead of model to filter

mikaelfr committed:
Pass datamodel instead of model to filter

mikaelfr committed:
Move filters to utils and add check for task tags in UtcAwareFilterConverter

mikaelfr committed:
Move filters to utils and add check for task tags in UtcAwareFilterConverter

mikaelfr committed:
Check class in is_tasktag only for relations

mikaelfr committed:
Check class in is_tasktag only for relations

mikaelfr committed:
Use custom ModelConverter

GeneralModelConverter tries to convert relationships first which overrides the field type we set for task tags in airflow wwwutils.

mikaelfr committed:
Use custom ModelConverter

GeneralModelConverter tries to convert relationships first which overrides the field type we set for task tags in airflow wwwutils.

mikaelfr committed:
Define custom TaskTagModelView which uses TaskTagModelConverter instead of GeneralModelConverter

mikaelfr committed:
Define custom TaskTagModelView which uses TaskTagModelConverter instead of GeneralModelConverter

mikaelfr committed:
Fix get_col_default crash on attempting to call a relationship

mikaelfr committed:
Fix get_col_default crash on attempting to call a relationship

mikaelfr committed:
Add example dag for task tags

mikaelfr committed:
Add example dag for task tags

mikaelfr committed:
Add tests for tag filter view

mikaelfr committed:
Add tests for tag filter view

mikaelfr committed:
Add on cascade delete to TaskTag relationship

mikaelfr committed:
Add on cascade delete to TaskTag relationship

mikaelfr committed:
Fix serialization tests

mikaelfr committed:
Fix serialization tests

mikaelfr committed:
Fix serialization

mikaelfr committed:
Fix serialization

mikaelfr committed:
Replace tags with _tags in serialization test

mikaelfr committed:
Replace tags with _tags in serialization test

mikaelfr committed:
Fix TaskTag on delete cascade not working

mikaelfr committed:
Fix TaskTag on delete cascade not working

mikaelfr committed:
Fix typo in TaskInstanceView tests

mikaelfr committed:
Fix typo in TaskInstanceView tests

mikaelfr committed:
Fix task instance db and model mismatch, merge heads

mikaelfr committed:
Fix task instance db and model mismatch, merge heads

mikaelfr committed:
Rename tags to task_tags since tags is already used in ECSOperator

mikaelfr committed:
Rename tags to task_tags since tags is already used in ECSOperator

mikaelfr committed:
Modify serialization to reflect renaming tags to task_tags

mikaelfr committed:
Modify serialization to reflect renaming tags to task_tags

mikaelfr committed:
Reflect renaming tags in test_views and example dag

mikaelfr committed:
Reflect renaming tags in test_views and example dag

mikaelfr committed:
Change tags to task_tags in FAB views and reflect changed endpoint in tests

mikaelfr committed:
Change tags to task_tags in FAB views and reflect changed endpoint in tests

mikaelfr committed:
Rename tags to task_tags in graph view

mikaelfr committed:
Rename tags to task_tags in graph view

mikaelfr committed:
Sort imports

mikaelfr committed:
Sort imports

mikaelfr committed:
Static code check fixes

mikaelfr committed:
Static code check fixes

mikaelfr committed:
Fix mypy errors in BaseOperator

mikaelfr committed:
Fix mypy errors in BaseOperator

mikaelfr committed:
Add docs and fix bug in views

* Add docs for adding task tags
* Update task tags example by adding an etl tag to some tasks
* Fix rename tags to task_tags on graph refresh

mikaelfr committed:
Add docs and fix bug in views

* Add docs for adding task tags
* Update task tags example by adding an etl tag to some tasks
* Fix rename tags to task_tags on graph refresh

mikaelfr committed:
Add docs for filtering tags

mikaelfr committed:
Add docs for filtering tags

mikaelfr committed:
Change filtering to be a subsection

mikaelfr committed:
Change filtering to be a subsection

ashb commented on airflow/migrations/versions/7c1bef9cb7b9_merge_heads.py:
```diff
@@ -0,0 +1,41 @@
+
+# revision identifiers, used by Alembic.
+revision = '7c1bef9cb7b9'
+down_revision = ('3c20cacc0044', '19d8a998c007')
```
There should not be a merge migration. Please rebase your change and update the `down_revision` on your new migratin instead.

mikaelfr replied:
Fixed in 648e9c2

ashb commented on airflow/models/baseoperator.py:
```diff
@@ -243,6 +243,8 @@ class derived from this one results in the creation of a task object,
     :param do_xcom_push: if True, an XCom is pushed containing the Operator's
         result
     :type do_xcom_push: bool
+    :param task_tags: List of tags used to identify this task
```
```suggestion
    :param tags: List of tags used to identify this task
```

We don't need to say "task" here -- it's already on a task.

mikaelfr replied:
It was originally just called `tags` but when some tests failed I found out that [ECSOperator](https://github.com/apache/airflow/blob/e2b219823b9435dcdb0956a64ee2fdbf2e475124/airflow/providers/amazon/aws/operators/ecs.py#L89-L90), [OpsgenieAlertOperator](https://github.com/apache/airflow/blob/e2b219823b9435dcdb0956a64ee2fdbf2e475124/airflow/providers/opsgenie/operators/opsgenie_alert.py#L50-L51) and [AzureContainerInstancesOperator](https://github.com/apache/airflow/blob/e2b219823b9435dcdb0956a64ee2fdbf2e475124/airflow/providers/microsoft/azure/operators/azure_container_instances.py#L90-L91) already have a `tags` parameter.

ashb commented on airflow/models/baseoperator.py:
```diff
@@ -359,6 +363,8 @@ def __init__(
         self.email_on_retry = email_on_retry
         self.email_on_failure = email_on_failure
+        self._task_tags: Optional[List[str]] = []
+        self.task_tags = task_tags or []
```
```suggestion
        self.task_tags = task_tags
```

The setter already handles None.

mikaelfr replied:
Fixed in e67fd78

ashb commented on airflow/models/dag.py:
```diff
@@ -844,11 +844,12 @@ def get_task_instances(
                 datetime.combine(start_date, datetime.min.time()))
 
-        tis = session.query(TaskInstance).filter(
+        tis = session.query(TaskInstance).options(joinedload(TaskInstance.task_tags)).filter(
```
Make this an optional flag on this method -- the scheduler doesn't care about tags so shouldn't pay the cost of loading them.

mikaelfr replied:
I took a look at SQLAlchemy [documentation](https://docs.sqlalchemy.org/en/13/orm/loading_relationships.html) and modified the relationship to raise an exception on lazy loads by default. I then added explicit `joinload`s in a few specific places where we want to load task tags. So currently the tags are not loaded by default and are only loaded with a join in a few places.

ashb commented on airflow/models/taskinstance.py:
```diff
@@ -501,6 +524,13 @@ def refresh_from_task(self, task, pool_override=None):
         self.max_tries = task.retries
         self.executor_config = task.executor_config
         self.operator = task.__class__.__name__
+        # Mocks tags attribute is not iterable
```
Don't put support code for tests in the real class -- provide a better mock instead.

mikaelfr replied:
Fixed in fb4f595

ashb commented on airflow/models/taskinstance.py:
```diff
@@ -501,6 +524,13 @@ def refresh_from_task(self, task, pool_override=None):
+                self.task_tags.append(
+                    TaskTag(name=tag, dag_id=self.dag_id,
+                            task_id=self.task_id, execution_date=self.execution_date)
+                )
```
Doesn't SQLA handle this automatically for the relationship?
```suggestion
                self.task_tags.append(TaskTag(name=tag))
```

mikaelfr replied:
Apparently it does. Fixed in 9e683c4

ashb reviewed (changes requested):
Please provide screenshots of the new UI functionality.

I'm not sure storing this against TI is actually the right thing to do -- the task tag is a property of the Task, not the TaskInstance. What about instead if we just looked at the tags property of `dag.tasks`.

i.e. something like `tasks = [task for dag.tasks if task.has_tag(val)]`

(With the planned "DAG versioning" where we store a different version of the serialized dag model when it changes any concern about this being out of date with latest definition goes away)

ashb commented on docs/howto/add-task-tags.rst:
```diff
@@ -0,0 +1,41 @@
+    under the License.
+
+
+Add tags to tasks
```
I think this should be in the add-dag-tags.rst -- a separate page makes it harder to discover this feature.

mikaelfr replied:
Fixed in 8c67348

ashb commented on airflow/example_dags/example_task_tags.py:
```diff
@@ -0,0 +1,63 @@
+#
```
This file is never used by anything -- please don't add new example dags without very good reason as each file in this folder slows down our test suite.

mikaelfr replied:
Fixed in 8c67348

mikaelfr committed:
Fix mysql timestamping and remove merge heads

mikaelfr committed:
Fix mysql timestamping and remove merge heads

mikaelfr committed:
Update docs and example dags

Remove unnecessary example dag. Combine how-to documentation for dag and task tags into a single page.

mikaelfr committed:
Update docs and example dags

Remove unnecessary example dag. Combine how-to documentation for dag and task tags into a single page.

mikaelfr committed:
Fix unnecessary params when adding TaskTags

Fix TaskInstance model and db mismatch in fk definition

mikaelfr committed:
Fix unnecessary params when adding TaskTags

Fix TaskInstance model and db mismatch in fk definition

mikaelfr committed:
Remove support code for tests from TaskInstance

* Remove check for task_tags iterability in TaskInstance
* Add task_tags empty lists to Task mocks

mikaelfr committed:
Remove support code for tests from TaskInstance

* Remove check for task_tags iterability in TaskInstance
* Add task_tags empty lists to Task mocks

mikaelfr committed:
Fix diverged head and static checks

mikaelfr committed:
Fix diverged head and static checks

mikaelfr committed:
Fix brain fart in tags docs

mikaelfr committed:
Remove redundant None-safe assignment

Ignoring type here since mypy doesn't understand differing getter and setter types as described in python/mypy#3004

mikaelfr committed:
Remove redundant None-safe assignment

Ignoring type here since mypy doesn't understand differing getter and setter types as described in python/mypy#3004

mikaelfr committed:
Use raiseload for TaskInstance.task_tags by default

This allows task_tags to be queried from TaskInstance only when that is explicitly desired.

mikaelfr committed:
Use raiseload for TaskInstance.task_tags by default

This allows task_tags to be queried from TaskInstance only when that is explicitly desired.

mikaelfr committed:
Fix static checks after rebase

mikaelfr committed:
Fix static checks after rebase

mikaelfr committed:
Add task_tag loading to refresh

mikaelfr committed:
Add task_tag loading to refresh

mikaelfr committed:
Fix bugs caused by rebase

mikaelfr committed:
Fix diverged heads and dag creation in tests

mikaelfr committed:
Fix errors caused by updated flask-appbuilder version
